### PR TITLE
update function calling api to tools api

### LINKED ISSUE #293: Upgrade: OpenAI function_call (deprecated) ==> tools

### PR DESCRIPTION
**Pull request by nilspalumbo**
Resolves #293; updates default Azure API version to latest, which supports tools.

### CODE REVIEW

**Comment by Mohannadcse** [comment]
@pchalasani the pytest has been running for almost 6 hrs, I think it should be cancelled

File: langroid/language_models/azure_openai.py [hunk 1]
```diff
@@ -29,7 +29,7 @@ class AzureConfig(OpenAIGPTConfig):
 
     api_key: str = ""  # CAUTION: set this ONLY via env var AZURE_OPENAI_API_KEY
     type: str = "azure"
-    api_version: str = "2023-05-15"
+    api_version: str = "2024-02-15-preview"
```
Comment (Mohannadcse) [inline langroid/language_models/azure_openai.py:32]: good to update here, can be overwritten via the `AZURE_OPENAI_API_VERSION` in the `.env`

**Comment by Mohannadcse** [comment]
@nilspalumbo shouldn't docstring be added/updated to clarify the purpose of the mods involved `| Dict[str, str]`

**Comment by nilspalumbo** [comment]
Oof, I hope OpenAI rate limited this one :grimacing: Though, when I look at it, `test_multi_agent.py` has four tests, one of which has two configs, and there are five successes; could the tasks be hanging on after the test completed?

**Comment by pchalasani** [comment]
@nilspalumbo `test_tool_messages.py` is failing

```
tests/main/test_tool_messages.py ........F

========================================================================== FAILURES ==========================================================================
___________________________________________________ test_enable_message[FileExistsMessage-True-True-True] ____________________________________________________

msg_class = <class 'test_tool_messages.FileExistsMessage'>, use = True, handle = True, force = True

    @pytest.mark.parametrize(
        # cartesian product of all combinations of use, handle, force
        "msg_class, use, handle, force",
        cartesian_product,
    )
    def test_enable_message(
        msg_class: Optional[ToolMessage], use: bool, handle: bool, force: bool
    ):
        agent.enable_message(msg_class, use=use, handle=handle, force=force)
        tools = agent._get_tool_list(msg_class)
        for tool in tools:
            assert tool in agent.llm_tools_map
            if msg_class is not None:
                assert agent.llm_tools_map[tool] == msg_class
                assert agent.llm_functions_map[tool] == msg_class.llm_function_schema()
            assert (tool in agent.llm_tools_handled) == handle
            assert (tool in agent.llm_tools_usable) == use
            assert (tool in agent.llm_functions_handled) == handle
            assert (tool in agent.llm_functions_usable) == use

        if msg_class is not None:
>           assert (
                agent.llm_function_force is not None
                and agent.llm_function_force["name"] == tools[0]
            ) == force
E           KeyError: 'name'

tests/main/test_tool_messages.py:129: KeyError
```

**Comment by nilspalumbo** [comment]
Whoops, that makes sense, I'll update the test to be in terms of the new forcing format.

**Comment by pchalasani** [comment]
@nilspalumbo  Can we close this since full OpenAI tools support was added in 0.8.0
https://github.com/langroid/langroid/releases/tag/0.7.0

**Comment by nilspalumbo** [comment]
That's good to hear! Let's close this then, 0.7 should supersede it.

**Pull request closed without merging**